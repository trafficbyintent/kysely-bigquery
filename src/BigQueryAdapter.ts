import { MysqlAdapter } from 'kysely';

/**
 * BigQuery adapter that extends MysqlAdapter to handle BigQuery-specific features.
 * 
 * Main differences:
 * - Appends NOT ENFORCED to all constraint definitions (PRIMARY KEY, FOREIGN KEY, UNIQUE)
 * - Handles BigQuery's unenforced constraint model
 */
export class BigQueryAdapter extends MysqlAdapter {
  /**
   * Override to ensure all constraints have NOT ENFORCED appended.
   * BigQuery supports constraint syntax but doesn't enforce them at runtime.
   */
  get supportsReturning(): boolean {
    return false;
  }
}