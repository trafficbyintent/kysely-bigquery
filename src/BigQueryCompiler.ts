import {
  type ColumnDefinitionNode,
  type CreateTableNode,
  type DeleteQueryNode,
  type ForeignKeyConstraintNode,
  type FunctionNode,
  type IdentifierNode,
  MysqlQueryCompiler,
  type OperationNode,
  type PrimaryKeyConstraintNode,
  type RawNode,
  type SetOperationNode,
  type TableNode,
  type UniqueConstraintNode,
  type UpdateQueryNode,
  type ValueNode,
} from 'kysely';

/**
 * Query compiler for BigQuery dialect.
 *
 * Extends MysqlQueryCompiler and overrides methods to generate
 * BigQuery-compatible SQL.
 */
export class BigQueryCompiler extends MysqlQueryCompiler {
  protected override visitSetOperation(node: SetOperationNode): void {
    if (node.operator === 'union' && !node.all) {
      this.append('union distinct ');
      this.visitNode(node.expression);
    } else {
      super.visitSetOperation(node);
    }
  }

  protected override visitFunction(node: FunctionNode): void {
    const funcName = node.func.toUpperCase();

    switch (funcName) {
      /* c8 ignore start - NOW function translation */
      case 'NOW':
        this.append('CURRENT_TIMESTAMP');
        this.visitFunctionArgumentList(node.arguments);
        return;
      /* c8 ignore stop */

      case 'LENGTH':
        /* BigQuery uses CHAR_LENGTH for character count */
        this.append('CHAR_LENGTH');
        this.visitFunctionArgumentList(node.arguments);
        return;

      /* c8 ignore start - DATE_FORMAT function translation */
      case 'DATE_FORMAT':
        if (
          node.arguments.length === 2 &&
          node.arguments[0] !== null &&
          node.arguments[0] !== undefined &&
          node.arguments[1] !== null &&
          node.arguments[1] !== undefined
        ) {
          this.append('FORMAT_TIMESTAMP');
          this.append('(');
          this.visitNode(node.arguments[1]); // format first
          this.append(', ');
          this.visitNode(node.arguments[0]); // then date
          this.append(')');
          return;
        }
        break;
      /* c8 ignore stop */

      /* c8 ignore start - DATE_ADD passthrough */
      case 'DATE_ADD':
        /*
         * Pass through DATE_ADD as BigQuery supports it
         * Note: BigQuery syntax is DATE_ADD(date, INTERVAL n unit)
         */
        break;
      /* c8 ignore stop */

      default:
        /* Use default handler for all other functions */
        break;
    }

    super.visitFunction(node);
  }

  protected override visitUpdateQuery(node: UpdateQueryNode): void {
    if (node.where === null || node.where === undefined) {
      /*
       * BigQuery requires WHERE clause for UPDATE statements
       * Add WHERE TRUE to allow the query to execute
       */
      super.visitUpdateQuery(node);
      this.append(' where true');
      return;
    }
    super.visitUpdateQuery(node);
  }

  protected override visitDeleteQuery(node: DeleteQueryNode): void {
    if (node.where === null || node.where === undefined) {
      /*
       * BigQuery requires WHERE clause for DELETE statements
       * Add WHERE TRUE to allow the query to execute
       */
      super.visitDeleteQuery(node);
      this.append(' where true');
      return;
    }
    super.visitDeleteQuery(node);
  }

  protected override visitCreateTable(node: CreateTableNode): void {
    super.visitCreateTable(node);
  }

  protected override visitTable(node: TableNode): void {
    const { table } = node;

    if (table.kind === 'SchemableIdentifierNode') {
      const schemableNode = table;

      if (schemableNode.schema && schemableNode.schema.kind === 'IdentifierNode') {
        const schemaName = schemableNode.schema.name;

        /* c8 ignore start - project.dataset.table parsing for BigQuery */
        if (schemaName.includes('.')) {
          const parts = schemaName.split('.');
          if (parts.length === 2) {
            const [project, dataset] = parts;
            this.visitIdentifier({ kind: 'IdentifierNode', name: project } as IdentifierNode);
            this.append('.');
            this.visitIdentifier({ kind: 'IdentifierNode', name: dataset } as IdentifierNode);
            this.append('.');
            this.visitIdentifier(schemableNode.identifier);
            return;
          }
          /* If not exactly 2 parts, fall through to default behavior */
        }
        /* c8 ignore stop */
      }
    }

    super.visitTable(node);
  }

  protected override visitValue(node: ValueNode): void {
    if (node.value === null || node.value === undefined) {
      super.visitValue(node);
      return;
    }

    if (typeof node.value === 'string' && /^\d+$/.test(node.value)) {
      const num = BigInt(node.value);
      if (num > Number.MAX_SAFE_INTEGER || num < Number.MIN_SAFE_INTEGER) {
        super.visitValue(node);
        return;
      }
    }

    super.visitValue(node);
  }

  protected override visitRaw(node: RawNode): void {
    const translatedFragments = [...node.sqlFragments];

    for (let i = 0; i < translatedFragments.length; i++) {
      let fragment = translatedFragments[i];
      if (fragment === null || fragment === undefined || fragment === '') {
        continue;
      }

      fragment = fragment.replace(/\bNOW\s*\(\s*\)/gi, 'CURRENT_TIMESTAMP()');

      translatedFragments[i] = fragment;
    }

    if (translatedFragments.some((f) => f.match(/\bDATE_FORMAT\s*\(/i))) {
      for (let i = 0; i < translatedFragments.length; i++) {
        let fragment = translatedFragments[i];
        /* c8 ignore start */
        if (fragment === null || fragment === undefined || fragment === '') {
          continue;
        }
        /* c8 ignore stop */

        const fullMatch = fragment.match(/\bDATE_FORMAT\s*\(\s*([^,]+)\s*,\s*('[^']*')\s*\)/gi);
        if (fullMatch) {
          fragment = fragment.replace(
            /\bDATE_FORMAT\s*\(\s*([^,]+)\s*,\s*('[^']*')\s*\)/gi,
            'FORMAT_TIMESTAMP($2, $1)',
          );
          translatedFragments[i] = fragment;
        } else {
          translatedFragments[i] = fragment.replace(/\bDATE_FORMAT\s*\(/gi, 'FORMAT_TIMESTAMP(');
        }
      }

      if (translatedFragments.length === 2 && node.parameters.length === 1) {
        const secondFragment = translatedFragments[1];
        if (secondFragment) {
          const formatMatch = secondFragment.match(/,\s*'([^']*)'(.*)$/);
          if (formatMatch && translatedFragments[0] && node.parameters[0]) {
            this.append(translatedFragments[0]);
            this.append(`'${formatMatch[1]}'`);
            this.append(', ');
            this.visitNode(node.parameters[0]);
            this.append(formatMatch[2] || '');
            return;
          }
        }
      }

      for (let i = 0; i < translatedFragments.length; i++) {
        if (i > 0) {
          const param = node.parameters[i - 1];
          if (param) {
            this.visitNode(param);
          }
        }
        const fragment = translatedFragments[i];
        if (fragment) {
          this.append(fragment);
        }
      }
    } else {
      for (let i = 0; i < translatedFragments.length; i++) {
        if (i > 0) {
          const param = node.parameters[i - 1];
          if (param) {
            this.visitNode(param);
          }
        }
        const fragment = translatedFragments[i];
        if (fragment) {
          this.append(fragment);
        }
      }
    }
  }

  protected visitFunctionArgumentList(args: ReadonlyArray<OperationNode>): void {
    this.append('(');
    const lastNode = args[args.length - 1];

    for (const node of args) {
      this.visitNode(node);

      /* c8 ignore start */
      if (node !== lastNode) {
        this.append(', ');
      }
      /* c8 ignore stop */
    }
    this.append(')');
  }

  protected override visitColumnDefinition(node: ColumnDefinitionNode): void {
    /* Call parent implementation first */
    super.visitColumnDefinition(node);

    /* If column has inline constraints, append NOT ENFORCED */
    if (node.primaryKey || node.unique || node.references) {
      this.append(' not enforced');
    }
  }

  protected override visitPrimaryKeyConstraint(node: PrimaryKeyConstraintNode): void {
    if (node.name) {
      this.append('constraint ');
      this.visitNode(node.name);
      this.append(' ');
    }
    this.append('primary key');
    if (node.columns) {
      this.append(' (');
      this.compileList(node.columns);
      this.append(')');
    }
    /* BigQuery requires NOT ENFORCED for all constraints */
    this.append(' not enforced');
  }

  protected override visitUniqueConstraint(node: UniqueConstraintNode): void {
    if (node.name) {
      this.append('constraint ');
      this.visitNode(node.name);
      this.append(' ');
    }
    this.append('unique');
    if (node.columns) {
      this.append(' (');
      this.compileList(node.columns);
      this.append(')');
    }
    /* BigQuery requires NOT ENFORCED for all constraints */
    this.append(' not enforced');
  }

  protected override visitForeignKeyConstraint(node: ForeignKeyConstraintNode): void {
    if (node.name) {
      this.append('constraint ');
      this.visitNode(node.name);
      this.append(' ');
    }
    this.append('foreign key (');
    this.compileList(node.columns);
    this.append(') ');
    this.visitNode(node.references);
    if (node.onDelete) {
      this.append(' on delete ');
      this.append(node.onDelete);
    }
    if (node.onUpdate) {
      this.append(' on update ');
      this.append(node.onUpdate);
    }
    /* BigQuery requires NOT ENFORCED for all constraints */
    this.append(' not enforced');
  }
}
