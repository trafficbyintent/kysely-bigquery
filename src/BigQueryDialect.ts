import { BigQuery, BigQueryOptions, Dataset, Table } from '@google-cloud/bigquery';
import {
  DatabaseIntrospector,
  Dialect,
  Driver,
  Kysely,
  QueryCompiler,
} from 'kysely';
import {BigQueryAdapter} from './bigQueryAdapter';
import {BigQueryDriver} from './bigQueryDriver';
import {BigQueryIntrospector} from './bigQueryIntrospector';
import {BigQueryCompiler} from './bigQueryCompiler';

/**
 * Configuration options for BigQuery dialect.
 */
export interface BigQueryDialectConfig {
  /**
   * BigQuery client options.
   */
  options?: BigQueryOptions;
  
  /**
   * Existing BigQuery, Dataset, or Table instance to use.
   * If provided, this will be used instead of creating a new client.
   */
  bigquery?: BigQuery | Dataset | Table;
  
  /**
   * Optional mapping of table names to their JSON column names.
   * This enables automatic JSON serialization for these columns.
   * 
   * Example:
   * ```
   * {
   *   'dataset.users': ['metadata', 'settings'],
   *   'dataset.products': ['specifications']
   * }
   * ```
   */
  jsonColumns?: Record<string, string[]>;
}

/**
 * BigQuery dialect for Kysely.
 * 
 * This dialect allows you to use Kysely with Google BigQuery.
 */
export class BigQueryDialect implements Dialect {
  readonly #config: BigQueryDialectConfig;

  constructor(config?: BigQueryDialectConfig) {
    this.#config = this.#validateConfig(config ?? {});
  }

  createAdapter() {
    return new BigQueryAdapter();
  }

  createDriver(): Driver {
    return new BigQueryDriver(this.#config);
  }

  createQueryCompiler(): QueryCompiler {
    return new BigQueryCompiler();
  }

  createIntrospector(db: Kysely<unknown>): DatabaseIntrospector {
    return new BigQueryIntrospector(db, this.#config);
  }

  #validateConfig(config: BigQueryDialectConfig): BigQueryDialectConfig {
    /* Validate mutually exclusive options */
    if (config.options && config.bigquery) {
      throw new Error(
        'Cannot provide both "options" and "bigquery" in BigQueryDialectConfig. ' +
        'Use either "options" to create a new client or "bigquery" to use an existing instance.',
      );
    }

    /* Validate BigQuery options if provided */
    if (config.options) {
      /* Ensure projectId is provided when using options */
      if (!config.options.projectId && !process.env['GOOGLE_CLOUD_PROJECT']) {
        throw new Error(
          'BigQuery projectId is required. Provide it in options.projectId or set GOOGLE_CLOUD_PROJECT environment variable.',
        );
      }

      /* Validate credentials */
      if (!config.options.keyFilename && !config.options.credentials && !process.env['GOOGLE_APPLICATION_CREDENTIALS']) {
        /* Authentication will fall back to Application Default Credentials */
      }
    }

    /* Validate bigquery instance if provided */
    if (config.bigquery) {
      const instance = config.bigquery;
      const hasQueryMethod = typeof (instance as any).query === 'function';
      const hasCreateQueryStreamMethod = typeof (instance as any).createQueryStream === 'function';
      
      if (!hasQueryMethod || !hasCreateQueryStreamMethod) {
        throw new Error(
          'Invalid bigquery instance provided. It must have query() and createQueryStream() methods.',
        );
      }
    }

    return config;
  }
}

