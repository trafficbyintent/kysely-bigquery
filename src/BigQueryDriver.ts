import {
  DatabaseConnection,
  Driver,
} from 'kysely';
import { BigQueryDialectConfig } from '.';
import { BigQueryConnection } from './BigQueryConnection';

export class BigQueryDriver implements Driver {
  #config: BigQueryDialectConfig;

  constructor(config: BigQueryDialectConfig) {
    this.#config = config;
  }

  async init(): Promise<void> {}

  async acquireConnection(): Promise<DatabaseConnection> {
    return new BigQueryConnection(this.#config);
  }

  async beginTransaction(conn: BigQueryConnection): Promise<void> {
    return await conn.beginTransaction();
  }

  async commitTransaction(conn: BigQueryConnection): Promise<void> {
    return await conn.commitTransaction();
  }

  async rollbackTransaction(conn: BigQueryConnection): Promise<void> {
    return await conn.rollbackTransaction();
  }

  async releaseConnection(_conn: BigQueryConnection): Promise<void> {}

  async destroy(): Promise<void> {}
}