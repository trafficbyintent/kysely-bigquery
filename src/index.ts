export * from './BigQueryConnection';
export * from './BigQueryDialect';
export * from './BigQueryDriver';
export * from './BigQueryIntrospector';