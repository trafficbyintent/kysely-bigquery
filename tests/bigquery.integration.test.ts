import {expect, test, beforeAll, afterAll} from 'vitest';
import {Kysely, sql} from 'kysely';

import {BigQueryDialect} from '../src';
import {createBigQueryInstance} from './config';
import {
  generateTestId,
  createTestTable,
  cleanupTestTable,
  TEST_USERS_SCHEMA,
  TEST_PRODUCTS_SCHEMA,
  TEST_ORDERS_SCHEMA,
} from './helpers';

const kysely = new Kysely<any>({
  dialect: new BigQueryDialect({
    bigquery: createBigQueryInstance(),
  }),
});

// Setup test tables before running tests
beforeAll(async () => {
  try {
    await createTestTable(kysely, 'test_dataset', 'test_users', TEST_USERS_SCHEMA);
    await createTestTable(kysely, 'test_dataset', 'test_products', TEST_PRODUCTS_SCHEMA);
    await createTestTable(kysely, 'test_dataset', 'test_orders', TEST_ORDERS_SCHEMA);
  } catch (error) {
    console.error('Failed to create test tables:', error);
    throw error;
  }
});

// Cleanup test tables after all tests
afterAll(async () => {
  try {
    await cleanupTestTable(kysely, 'test_dataset', 'test_users');
    await cleanupTestTable(kysely, 'test_dataset', 'test_products');
    await cleanupTestTable(kysely, 'test_dataset', 'test_orders');
  } catch (error) {
    console.error('Failed to cleanup test tables:', error);
  }
});

test('simple select execution', async () => {
  const query = kysely.selectFrom('features.metadata').where('id', '>', 10).selectAll().limit(1);

  const rows = await query.execute();

  expect(rows).toHaveLength(1);
  
  // Since the result has BigQueryTimestamp objects, we'll just verify the structure
  const row = rows[0];
  expect(row.id).toBe(28);
  expect(row.category).toBe('cac');
  expect(row.name).toBe('cac');
  expect(row.created_at).toHaveProperty('value');
  expect(row.updated_at).toHaveProperty('value');
  expect(row.inserted_at).toHaveProperty('value');
});

test('introspection', async () => {
  const tables = await kysely.introspection.getTables();

  expect(tables.filter(t => t.name === 'bank_account_transactions')).toMatchSnapshot();
});

test('INSERT operations', async () => {
  // Test basic insert
  const insertResult = await kysely
    .insertInto('test_dataset.test_users')
    .values({
      id: generateTestId(),
      name: 'Test User',
      email: 'test@example.com',
      age: 25,
      tags: ['test', 'integration'],
      metadata: {
        source: 'test_suite',
        verified: true,
      },
      created_at: new Date(),
    })
    .execute();

  // BigQuery doesn't return affected rows, but the query should execute without error
  expect(insertResult).toBeDefined();

  // Test multiple inserts
  const testId1 = generateTestId();
  const testId2 = generateTestId();
  
  await kysely
    .insertInto('test_dataset.test_users')
    .values([
      {
        id: testId1,
        name: 'User 1',
        email: 'user1@example.com',
        created_at: new Date(),
      },
      {
        id: testId2,
        name: 'User 2',
        email: 'user2@example.com',
        created_at: new Date(),
      },
    ])
    .execute();

  // Verify the inserts worked
  const users = await kysely
    .selectFrom('test_dataset.test_users')
    .where('id', 'in', [testId1, testId2])
    .selectAll()
    .execute();

  expect(users).toHaveLength(2);
  expect(users.map(u => u.name).sort()).toEqual(['User 1', 'User 2']);
});

test('UPDATE operations', async () => {
  // First insert a test record
  const testId = generateTestId();
  await kysely
    .insertInto('test_dataset.test_users')
    .values({
      id: testId,
      name: 'Original Name',
      email: 'original@example.com',
      age: 30,
      created_at: new Date(),
    })
    .execute();

  // Update the record
  await kysely
    .updateTable('test_dataset.test_users')
    .set({
      name: 'Updated Name',
      email: 'updated@example.com',
      age: 31,
      updated_at: new Date(),
    })
    .where('id', '=', testId)
    .execute();

  // Verify the update
  const updated = await kysely
    .selectFrom('test_dataset.test_users')
    .where('id', '=', testId)
    .selectAll()
    .executeTakeFirst();

  expect(updated).toBeDefined();
  expect(updated!.name).toBe('Updated Name');
  expect(updated!.email).toBe('updated@example.com');
  expect(updated!.age).toBe(31);
  expect(updated!.updated_at).toBeDefined();
});

test('DELETE operations', { timeout: 30000 }, async () => {
  // Insert test records
  const testIds = [generateTestId(), generateTestId(), generateTestId()];
  
  await kysely
    .insertInto('test_dataset.test_users')
    .values(testIds.map((id, index) => ({
      id,
      name: `Delete Test ${index}`,
      email: `delete${index}@example.com`,
      created_at: new Date(),
    })))
    .execute();

  // Delete one record
  await kysely
    .deleteFrom('test_dataset.test_users')
    .where('id', '=', testIds[0])
    .execute();

  // Delete multiple records with condition
  await kysely
    .deleteFrom('test_dataset.test_users')
    .where('id', 'in', [testIds[1], testIds[2]])
    .execute();

  // Verify all are deleted
  const remaining = await kysely
    .selectFrom('test_dataset.test_users')
    .where('id', 'in', testIds)
    .selectAll()
    .execute();

  expect(remaining).toHaveLength(0);
});

test('Complex data modification scenarios', { timeout: 30000 }, async () => {
  // Test INSERT with subquery
  const sourceId = generateTestId();
  const targetId = generateTestId();
  
  // Insert source data
  await kysely
    .insertInto('test_dataset.test_products')
    .values({
      id: sourceId,
      name: 'Source Product',
      price: 99.99,
      category: 'electronics',
      tags: ['source', 'test'],
      created_at: new Date(),
    })
    .execute();

  // Insert with SELECT subquery
  await sql`
    INSERT INTO test_dataset.test_orders (id, customer_id, product_id, quantity, total_amount, status, order_date, created_at)
    SELECT 
      ${targetId},
      'customer_123',
      id,
      2,
      price * 2,
      'pending',
      CURRENT_DATE(),
      CURRENT_TIMESTAMP()
    FROM test_dataset.test_products
    WHERE id = ${sourceId}
  `.execute(kysely);

  // Verify the insert
  const order = await kysely
    .selectFrom('test_dataset.test_orders')
    .where('id', '=', targetId)
    .selectAll()
    .executeTakeFirst();

  expect(order).toBeDefined();
  expect(order!.product_id).toBe(sourceId);
  expect(order!.quantity).toBe(2);
  expect(order!.total_amount).toBe(199.98);

  // Test UPDATE with JOIN (BigQuery syntax)
  await sql`
    UPDATE test_dataset.test_orders o
    SET o.status = 'shipped',
        o.shipped_date = CURRENT_DATE()
    FROM test_dataset.test_products p
    WHERE o.product_id = p.id
    AND p.category = 'electronics'
    AND o.id = ${targetId}
  `.execute(kysely);

  // Verify the update
  const updatedOrder = await kysely
    .selectFrom('test_dataset.test_orders')
    .where('id', '=', targetId)
    .selectAll()
    .executeTakeFirst();

  expect(updatedOrder!.status).toBe('shipped');
  expect(updatedOrder!.shipped_date).toBeDefined();
});

test('BigQuery ARRAY type handling', async () => {
  const testId = generateTestId();
  
  // Insert data with ARRAY
  await kysely
    .insertInto('test_dataset.test_users')
    .values({
      id: testId,
      name: 'Array Test User',
      email: 'array@example.com',
      tags: ['developer', 'typescript', 'bigquery'],
      created_at: new Date(),
    })
    .execute();

  // Query with ARRAY operations
  const result = await kysely
    .selectFrom('test_dataset.test_users')
    .select(['id', 'name', 'tags'])
    .where('id', '=', testId)
    .executeTakeFirst();

  expect(result).toBeDefined();
  expect(result!.tags).toEqual(['developer', 'typescript', 'bigquery']);

  // Test ARRAY contains using raw SQL
  const hasTag = await sql`
    SELECT EXISTS(
      SELECT 1 
      FROM test_dataset.test_users 
      WHERE id = ${testId}
      AND 'typescript' IN UNNEST(tags)
    ) as has_tag
  `.execute(kysely);

  expect(hasTag.rows[0].has_tag).toBe(true);

  // Test ARRAY_LENGTH
  const arrayLength = await sql`
    SELECT ARRAY_LENGTH(tags) as tag_count
    FROM test_dataset.test_users
    WHERE id = ${testId}
  `.execute(kysely);

  expect(arrayLength.rows[0].tag_count).toBe(3);
});

test('BigQuery STRUCT type handling', async () => {
  const testId = generateTestId();
  
  // Insert data with STRUCT
  await kysely
    .insertInto('test_dataset.test_users')
    .values({
      id: testId,
      name: 'Struct Test User',
      email: 'struct@example.com',
      metadata: {
        source: 'integration_test',
        verified: true,
      },
      created_at: new Date(),
    })
    .execute();

  // Query STRUCT data
  const result = await kysely
    .selectFrom('test_dataset.test_users')
    .select(['id', 'name', 'metadata'])
    .where('id', '=', testId)
    .executeTakeFirst();

  expect(result).toBeDefined();
  expect(result!.metadata).toEqual({
    source: 'integration_test',
    verified: true,
  });

  // Query specific STRUCT fields
  const structField = await sql`
    SELECT 
      metadata.source as source,
      metadata.verified as verified
    FROM test_dataset.test_users
    WHERE id = ${testId}
  `.execute(kysely);

  expect(structField.rows[0].source).toBe('integration_test');
  expect(structField.rows[0].verified).toBe(true);
});

test('BigQuery nested STRUCT handling', async () => {
  const testId = generateTestId();
  
  // Insert product with nested STRUCT
  await kysely
    .insertInto('test_dataset.test_products')
    .values({
      id: testId,
      name: 'Nested Struct Product',
      price: 299.99,
      details: {
        manufacturer: 'Test Corp',
        weight_kg: 2.5,
        dimensions: {
          length_cm: 30,
          width_cm: 20,
          height_cm: 10,
        },
      },
      created_at: new Date(),
    })
    .execute();

  // Query nested STRUCT
  const result = await sql`
    SELECT 
      details.manufacturer,
      details.weight_kg,
      details.dimensions.length_cm as length,
      details.dimensions.width_cm as width,
      details.dimensions.height_cm as height
    FROM test_dataset.test_products
    WHERE id = ${testId}
  `.execute(kysely);

  const row = result.rows[0];
  expect(row.manufacturer).toBe('Test Corp');
  expect(row.weight_kg).toBe(2.5);
  expect(row.length).toBe(30);
  expect(row.width).toBe(20);
  expect(row.height).toBe(10);
});

test('BigQuery DATE and TIMESTAMP types', async () => {
  const testId = generateTestId();
  const orderDate = '2024-01-15';  // BigQuery DATE needs string format
  const shippedDate = '2024-01-17';
  
  // Insert with DATE and TIMESTAMP
  await kysely
    .insertInto('test_dataset.test_orders')
    .values({
      id: testId,
      customer_id: 'cust_123',
      product_id: 'prod_456',
      quantity: 1,
      total_amount: 99.99,
      status: 'shipped',
      order_date: orderDate,
      shipped_date: shippedDate,
      created_at: new Date(),
    })
    .execute();

  // Query and verify date handling
  const result = await kysely
    .selectFrom('test_dataset.test_orders')
    .selectAll()
    .where('id', '=', testId)
    .executeTakeFirst();

  expect(result).toBeDefined();
  
  // BigQuery returns dates as objects with value property
  expect(result!.order_date).toBeDefined();
  expect(result!.shipped_date).toBeDefined();
  expect(result!.created_at).toBeDefined();

  // Test date functions
  const dateDiff = await sql`
    SELECT DATE_DIFF(shipped_date, order_date, DAY) as days_to_ship
    FROM test_dataset.test_orders
    WHERE id = ${testId}
  `.execute(kysely);

  expect(dateDiff.rows[0].days_to_ship).toBe(2);
});

test('BigQuery NUMERIC and BOOL types', async () => {
  const testId = generateTestId();
  
  // Test NUMERIC precision
  await kysely
    .insertInto('test_dataset.test_products')
    .values({
      id: testId,
      name: 'Precision Test Product',
      price: 1234.56, // NUMERIC(10, 2)
      in_stock: true,
      created_at: new Date(),
    })
    .execute();

  const result = await kysely
    .selectFrom('test_dataset.test_products')
    .select(['price', 'in_stock'])
    .where('id', '=', testId)
    .executeTakeFirst();

  expect(result).toBeDefined();
  expect(result!.price).toBe(1234.56);
  expect(result!.in_stock).toBe(true);

  // Test BOOL in WHERE clause
  const inStockProducts = await kysely
    .selectFrom('test_dataset.test_products')
    .select('id')
    .where('in_stock', '=', true)
    .where('id', '=', testId)
    .execute();

  expect(inStockProducts).toHaveLength(1);
});

test('BigQuery JSON type handling', async () => {
  const testId = generateTestId();
  const metadata = {
    tags: ['urgent', 'international'],
    shipping: {
      carrier: 'FedEx',
      tracking: '1234567890',
    },
    notes: 'Handle with care',
  };
  
  // Insert JSON data
  await sql`
    INSERT INTO test_dataset.test_orders (
      id, customer_id, product_id, quantity, total_amount, 
      status, order_date, metadata, created_at
    )
    VALUES (
      ${testId}, 'cust_json', 'prod_json', 1, 50.00,
      'pending', CURRENT_DATE(), 
      PARSE_JSON(${JSON.stringify(metadata)}),
      CURRENT_TIMESTAMP()
    )
  `.execute(kysely);

  // Query JSON data
  const result = await sql`
    SELECT 
      JSON_VALUE(metadata, '$.notes') as notes,
      JSON_VALUE(metadata, '$.shipping.carrier') as carrier,
      JSON_QUERY(metadata, '$.tags') as tags_json
    FROM test_dataset.test_orders
    WHERE id = ${testId}
  `.execute(kysely);

  const row = result.rows[0];
  expect(row.notes).toBe('Handle with care');
  expect(row.carrier).toBe('FedEx');
  expect(JSON.parse(row.tags_json)).toEqual(['urgent', 'international']);
});