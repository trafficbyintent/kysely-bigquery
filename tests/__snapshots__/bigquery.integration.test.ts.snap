// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`introspection 1`] = `
[
  {
    "columns": [
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "id",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "bank_account_id",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "remote_id",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "fpath_account_id",
      },
      {
        "dataType": "DATE",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "date",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "original_description",
      },
      {
        "dataType": "INT64",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "pending",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "processing_status",
      },
      {
        "dataType": "INT64",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "amount",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "iso_currency_code",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "unofficial_currency_code",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "bank_account_category_id",
      },
      {
        "dataType": "DATE",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "authorized_date",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "name",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "merchant_name",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "pending_transaction_id",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "transaction_type",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "account_owner",
      },
      {
        "dataType": "INT64",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "amount_usd",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "transaction_category",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "transaction_status",
      },
      {
        "dataType": "STRUCT<type STRING, name STRING>",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "counter_party",
      },
      {
        "dataType": "TIMESTAMP",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "created_at",
      },
      {
        "dataType": "TIMESTAMP",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "updated_at",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "ntropy_batch_id",
      },
      {
        "dataType": "TIMESTAMP",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enriched_at",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enriched_by",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enrichment_cat1",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enrichment_cat2",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enrichment_cat3",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enrichment_cat4",
      },
      {
        "dataType": "STRING",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "enrichment_merchant_name",
      },
      {
        "dataType": "TIMESTAMP",
        "hasDefaultValue": false,
        "isAutoIncrementing": false,
        "isNullable": true,
        "name": "inserted_at",
      },
    ],
    "isView": false,
    "name": "bank_account_transactions",
    "schema": "api",
  },
]
`;
